_posts/2022-01-21-safe-releases.md

```diff
@@ -0,0 +1,95 @@
+
+## 1-2-3: A slow, safe release process
+
+For the next piece, we'll take a simple scenario and describe the releases necessary for a safe migration. The request is: move the `email` column from the `user` table to the `user_details` table. Keep in mind that there are existing emails in the original table.
+
+My rule of thumb is: The more explicit, smaller steps, the safer the release (to a point!). The more you break things down into parts, the less likely any release will result in a broken build or put your application into a messy state that requires manual intervention.
+
+Using that rule, here's a well-organized plan to release this change:
+
+### Release 1 - Prepare your new column, begin data synchronization, and asynchronously prepare all data
+
+First and foremost, you'll want to create your new `email` column on the `user_details` table. If you were to run only this, you would have no data in the new column and therefore would not be able to use it yet with your application code.
+
+To ensure data is fully synched by the time the second release comes along, you'll want two things:
+Modify any code that previously set `email` on the `user` table to also update the `user_details` table. By updating both, you will make sure that any updates after your population scripts are reflected in the second table.
+An asynchronous task to backpopulate emails in the new table. Based on our additional considerations, we should not be populating large amounts of data in the migration itself. So a typical approach here would be to ship a rake task or other script that can be run on the server once the release goes out.
+
+Once this release is deployed, you can execute your data population task/script. Since it lives outside of a migration, you can handle any long-running tasks by splitting work up or putting it in a background queue. Once the data population is done, you are done with the first release and you DB is in a state where it is ready to receive calls from new code
+
+### Release 2 - Updates to the application code
+
+Next, you make the code changes. Anywhere that was previously calling `user.email`, you can now change to `user.user_details.email`. All users should have their emails populated in the new column, and any updates that happened after your data population script should have been duplicated from your code changes.
+
+At this point, you can also remove your glue code which was updating both `user.email` and `user_details.email`. It only needs to update the new column since your new code will only be using the new column.
+
+### Release 3 - Dropping old columns, code, and cleanup
+
+At this point, your application should be running in the way you want it to. You added a new column and populated it (via 2 mechanisms), and then released code changes to use your new column.
+
+It is now safe to drop the old `user.email` column since code is no longer using it.
+
+## Fin
```

Comment:
Nice summary. If you want to call out the point I made earlier about more releases and more steps doesn't necessarily mean more work. CI/CD takes care of the repetative steps, while as a developer, you can use a set of steps to ensure safe releases and reliablity of your applications even when you have big data migrations as a part of a release.